### PR TITLE
board/arduino-leonardo: add initial support (without usb cdc)

### PR DESCRIPTION
**Pull request by tprrt**
board/arduino-leonardo: add initial support (without usb cdc)

### CODE REVIEW

**Comment by aabadie** [comment]
Awesome!
briefly looking at the code, looks quite good already.
Thanks for providing this and welcome to RIOT !

**Comment by miri64** [comment]
For reference: I've got the board, so when it's time to test I'm happy to do so.

**Comment by miri64** [comment]
I get only

```
Connecting to programmer: .avrdude: butterfly_recv(): programmer is not responding

avrdude: butterfly_recv(): programmer is not responding
avrdude: butterfly_recv(): programmer is not responding
avrdude: butterfly_recv(): programmer is not responding
avrdude: butterfly_recv(): programmer is not responding
avrdude: butterfly_recv(): programmer is not responding
Found programmer: Id = ""; type = 
    Software Version = .; Hardware Version = .
avrdude: butterfly_recv(): programmer is not responding
avrdude: butterfly_recv(): programmer is not responding
avrdude: error: buffered memory access not supported. Maybe it isn't
a butterfly/AVR109 but a AVR910 device?
avrdude: initialization failed, rc=-1
avrdude: AVR device initialized and ready to accept instructions
avrdude: Device signature = 0x88fb48
avrdude: Expected signature for ATmega32U4 is 1E 95 87
avrdude: butterfly_recv(): programmer is not responding
avrdude: error: programmer did not respond to command: leave prog mode
avrdude: butterfly_recv(): programmer is not responding
avrdude: error: programmer did not respond to command: exit bootloader

avrdude done.  Thank you.

/home/mlenders/Repositories/RIOT-OS/RIOT/examples/hello-world/../../Makefile.include:368: recipe for target 'flash' failed
make: *** [flash] Error 1
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/examples/hello-world'
```

when trying to flash.

Could you maybe also update the udev-rules in `dist/tools/vagrant/udev_rules` for our vagrant users?

**Comment by miri64** [comment]
Needs rebase, and please address my comment.

**Comment by miri64** [comment]
Ping?

**Comment by tprrt** [comment]
I rebase the branch on master latest, in order to correct the conflicts.

**Comment by tprrt** [comment]
@miri64  I rebase this branch on master latest, in order to correct the conflicts.

**Comment by tprrt** [comment]
@aabadie, @miri64 : It is good for you so that this is merged in the next release ?

File: boards/common/arduino-atmega/board.c [hunk 1]
```diff
@@ -49,6 +56,12 @@ void board_init(void)
     /* initialize the board LED */
     gpio_init(LED0_PIN, GPIO_OUT);
     LED0_OFF;
+#ifdef CPU_ATMEGA32U4
```
Comment (aabadie) [inline boards/common/arduino-atmega/board.c]: IIUC, LED1 and LED2 are specific to the leonardo board but not to ATMEGA32U4 cpu. There are other arduino boards with this CPU but no on-board LED1/LED2 I think.
What about a dedicated board.c for leonardo ?

File: boards/common/arduino-atmega/include/board_common.h [hunk 2]
```diff
@@ -47,14 +49,35 @@ extern "C" {
 #define LED0_MASK           (1 << DDB5)
 #endif
 
+#ifdef CPU_ATMEGA32U4
+#define LED0_PIN            GPIO_PIN(2, 7) /*< BUILTIN LED */
```
Comment (aabadie) [inline boards/common/arduino-atmega/include/board_common.h]: The doxygen comment should be `/**< ... */`, here and below

File: cpu/atmega32u4/startup.c [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ */
+
+/**
+ * @ingroup     cpu_atmega32u4
+ * @{
+ *
+ * @file
+ * @brief       Startup code and interrupt vector definition
+ *
+ * @author     Thomas Perrot <thomas.perrot@tupi.fr>
```
Comment (aabadie) [inline cpu/atmega32u4/startup.c]: Nit: author name could be aligned with brief description above

File: cpu/atmega32u4/include/cpu_conf.h [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+#endif
+
+/**
+ * @name Kernel configuration
+ *
+ * Since printf seems to get memory allocated by the linker/avr-libc the stack
+ * size tested sucessfully even with pretty small stacks.k
+ * @{
+ */
+#define THREAD_EXTRA_STACKSIZE_PRINTF    (128)
```
Comment (aabadie) [inline cpu/atmega32u4/include/cpu_conf.h]: Use (128**U**), here a below

File: cpu/atmega32u4/doc.txt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+/**
+ * @defgroup        cpu_atmega32u4 Atmel ATmega32u4
+ * @ingroup         cpu
```
Comment (aabadie) [inline cpu/atmega32u4/doc.txt]: I would put it in `cpu_atmega_common`

File: sys/pipe/pipe_dynamic.c [hunk 6]
```diff
@@ -25,7 +25,7 @@
  * @}
  */
 
-#if defined(MCU_ATMEGA2560) || defined(MCU_ATMEGA1281) || defined(MCU_ATMEGA328P)
+#if defined(MCU_ATMEGA2560) || defined(MCU_ATMEGA1281) || defined(MCU_ATMEGA328P) || defined(MCU_ATMEGA32U4)
```
Comment (aabadie) [inline sys/pipe/pipe_dynamic.c]: The line could be split with \\

**Comment by miri64** [comment]
> @aabadie, @miri64 : It is good for you so that this is merged in the next release ?

I'll test if the issues mentioned earlier still prevail as soon as I'm back from my vacation (on Thursday)

> Could you maybe also update the udev-rules in dist/tools/vagrant/udev_rules for our vagrant users?

What about this? As far as I can see this was never addressed.

**Comment by miri64** [comment]
@tprrt I still get the same error when trying to flash...:

```
$ BOARD=arduino-leonardo make -C examples/hello-world/ flash
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/examples/hello-world'
Building application "hello-world" for "arduino-leonardo" with MCU "atmega32u4".

"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/boards/arduino-leonardo
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/boards/common/arduino-atmega
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/core
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/cpu/atmega32u4
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/cpu/atmega_common
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/cpu/atmega_common/periph
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/drivers
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/drivers/periph_common
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/auto_init
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/isrpipe
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/tsrb
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/uart_stdio
   text	   data	    bss	    dec	    hex	filename
   5344	    306	    711	   6361	   18d9	/home/mlenders/Repositories/RIOT-OS/RIOT/examples/hello-world/bin/arduino-leonardo/hello-world.elf
avrdude -p atmega32u4 -c avr109 -P /dev/ttyACM0 -b 115200 -F -D -U flash:w:bin/arduino-leonardo/hello-world.hex

Connecting to programmer: .avrdude: butterfly_recv(): programmer is not responding

avrdude: butterfly_recv(): programmer is not responding
avrdude: butterfly_recv(): programmer is not responding
avrdude: butterfly_recv(): programmer is not responding
avrdude: butterfly_recv(): programmer is not responding
avrdude: butterfly_recv(): programmer is not responding
Found programmer: Id = ""; type = 
    Software Version = .; Hardware Version = .
avrdude: butterfly_recv(): programmer is not responding
avrdude: butterfly_recv(): programmer is not responding
avrdude: error: buffered memory access not supported. Maybe it isn't
a butterfly/AVR109 but a AVR910 device?
avrdude: initialization failed, rc=-1
avrdude: AVR device initialized and ready to accept instructions
avrdude: Device signature = 0x88bb45
avrdude: Expected signature for ATmega32U4 is 1E 95 87
avrdude: butterfly_recv(): programmer is not responding
avrdude: error: programmer did not respond to command: leave prog mode
avrdude: butterfly_recv(): programmer is not responding
avrdude: error: programmer did not respond to command: exit bootloader

avrdude done.  Thank you.

/home/mlenders/Repositories/RIOT-OS/RIOT/examples/hello-world/../../Makefile.include:390: recipe for target 'flash' failed
make: *** [flash] Error 1
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/examples/hello-world'
```

**Comment by tprrt** [comment]
Thank you for your feedback, I will try to find free time to look at the points above, as soon as possible.

**Comment by miri64** [comment]
Ping @tprrt?

**Comment by tcschmidt** [comment]
@tprrt Want this continued?

**Comment by tprrt** [comment]
I think it's good now, all your remarks have been taken into account.

**Comment by tprrt** [comment]
@tcschmidt I would like, I took into account the remarks, I hope it will be good.

**Comment by tcschmidt** [comment]
@tprrt great. Let's see what @miri64 responds.

File: cpu/atmega32u4/include/cpu_conf.h (hunk 4 above)
Comment (tprrt) [reply cpu/atmega32u4/include/cpu_conf.h]: If I do that, I get compile errors:
/home/tperrot/dev/tprrt/RIOT/core/include/debug.h:55:79: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
         if ((sched_active_thread == NULL) || (sched_active_thread->stack_size >= THREAD_EXTRA_STACKSIZE_PRINTF)) { \

File: boards/common/arduino-atmega/include/board_common.h (hunk 2 above)
Comment (tprrt) [reply boards/common/arduino-atmega/include/board_common.h]: Done

File: cpu/atmega32u4/doc.txt (hunk 5 above)
Comment (tprrt) [reply cpu/atmega32u4/doc.txt]: Done

File: sys/pipe/pipe_dynamic.c (hunk 6 above)
Comment (tprrt) [reply sys/pipe/pipe_dynamic.c]: Done

**Comment by miri64** [comment]
Hi, I just tried to compile, but there seems to be a different issue now :frowning_face:... Since 098770aeda582 all atmega platforms seem to require some sort of compatibility linker script. @cladmi, @kYc0o maybe you can help?

**Comment by cladmi** [comment]
@miri64 They can be extracted from `avr-gcc -Wl,--verbose -mmcu=$(CPU)` so here `avr-gcc -Wl,--verbose -mmcu=atmega32u4` I put the command in the ldscripts commit message (with an extra `-` before `-mmcu` it seams…) but I could have put it before the include line too indeed :D 

https://gist.github.com/cladmi/fc2302847a4faf448322157aa7084856

**Comment by miri64** [comment]
@ZetaR60 @maribu maybe you can have a look?

**Comment by kYc0o** [comment]
I'll take a look and give feedback.

**Comment by tcschmidt** [comment]
Ping @miri64

**Comment by kYc0o** [comment]
I have now the board but couldn't test properly, just tried to adapt to #9130 though it doesnt change the behaviour.

**Comment by tprrt** [comment]
@miri64 That's very common problem referenced on Arduino forums, when you try to upload firmware to Arduino Leonardo and clones with avrdude, you get programmer timeout error message.

Reason for it is very simple - Arduino Leonardo should be reset to provide another serial port for flashing.

It is necessary to reset, the board just before to execute the flash command so that it is in the bootloader.

**Comment by maribu** [comment]
> @ZetaR60 @maribu maybe you can have a look?

Sorry for the late reply. I'll have a look at Tuesday. By then I should have compatible hardware.

**Comment by miri64** [comment]
@tprrt I'll have a look tomorrow. My leonardo is currently at the office.

**Comment by miri64** [comment]
Ok, I was now able to flash the leonardo when hitting the reset button just before flashing (I still needed to add the linker script manually). However, now I loose the `tty` after a few seconds (presumably after the bootloader exits).

**Comment by miri64** [comment]
(`serial.serialutil.SerialException: [Errno 2] could not open port /dev/ttyACM0: [Errno 2] No such file or directory: '/dev/ttyACM0'`)

```
$ make list-ttys
# hit reset
$ make list-ttys
/sys/bus/usb/devices/1-1.2.4: Arduino LLC Arduino Leonardo serial: '', tty(s): ttyACM0
$ sleep 5; make list-ttys
```

**Comment by maribu** [comment]
@miri64: The Leonardo does not have an integrated USB-TTL converter, so RIOT would need to implement it in software if we would want to access the shell without a dedicated USB-TTL adapter. But I'm 

@tprrt: Using a dedicated USB-TTL converter that I connected to the RX/TX pins (D0/D1) I still could not get any output. Could you provide some hints? It would be nice if you could extend `boards/arduino-leonardo/doc.txt` to contain that hint and the one regarding the flashing. On slow PCs this might be required to flash successfully:

 1. Run `make BOARD=arduino-leonardo`
 2. Type `make BOARD=arduino-lenoardo flash-only` but do not press Enter yet
 3. Reset the board
 4. Hit enter to confirm the `make BOARD=arduino-lenoardo flash-only` command

(`make BOARD=arduino-lenoardo flash` could take longer than the bootloader waits on slow PCs.)

**Comment by tprrt** [comment]
@maribu I will update the leonardo/doc.txt to explain how to use a dedicated usb-ttl converter.

**Comment by miri64** [comment]
> @maribu I will update the leonardo/doc.txt to explain how to use a dedicated usb-ttl converter.

Also please about the required reset before flashing.

**Comment by tprrt** [comment]
@miri64, @maribu I add the explanations for the reset, the ttl and moved the initialization of leds in led_init.c

**Comment by miri64** [comment]
OK! With pressing reset just before flashing and using a separate FTDI adapter I was able to finally flash and read the output of the `hello-world` example :-). Let's see what Murdock has to report.

**Comment by miri64** [comment]
@aabadie your comments were addressed long ago. Can your review be dismissed?

File: boards/common/arduino-atmega/include/periph_conf.h [hunk 7]
```diff
@@ -61,6 +63,17 @@ extern "C" {
 #define TIMER_0_ISRB        TIMER1_COMPB_vect
 #endif
 
+#ifdef CPU_ATMEGA32U4
+#define TIMER_NUMOF         (2U)
```
Comment (miri64) [inline boards/common/arduino-atmega/include/periph_conf.h]: `TIMER_CHANNELS` is missing (most errors I've seen so far on Murdock can be traced to this).

Comment (miri64) [reply boards/common/arduino-atmega/include/periph_conf.h]: (also you state here that `TIMER_NUMOF` is `2U`, but there is only the config for one timer provided).

File: boards/common/arduino-atmega/include/periph_conf.h [hunk 8]
```diff
@@ -61,6 +63,17 @@ extern "C" {
...
+
+#define TIMER_0             MEGA_TIMER1
+#define TIMER_0_MASK        &TIMSK1
+#define TIMER_0_FLAG        &TIFR1
+#define TIMER_0_ISRA        TIMER1_COMPA_vect
+#define TIMER_0_ISRB        TIMER1_COMPB_vect
```
Comment (miri64) [inline boards/common/arduino-atmega/include/periph_conf.h]: I'm not sure these are the right config parameters. `tests/periph_timer` fails to initialize `TIMER_0`.

File: boards/common/arduino-atmega/include/periph_conf.h (hunk 7 above)
Comment (tprrt) [reply boards/common/arduino-atmega/include/periph_conf.h]: I pushed a fix for that.

**Comment by miri64** [comment]
Now `tests/periph_timer` and `tests/periph_adc` are compiling. However, the `periph_timer` application is failing:

```
2018-12-18 15:30:37,687 - INFO # main(): This is RIOT! (Version: 2018.10-RC1-436-gdd6ec-HEAD)
2018-12-18 15:30:37,688 - INFO # 
2018-12-18 15:30:37,716 - INFO # Test for peripheral TIMERs
2018-12-18 15:30:37,717 - INFO # 
2018-12-18 15:30:37,738 - INFO # Available timers: 1
2018-12-18 15:30:37,739 - INFO # 
2018-12-18 15:30:37,757 - INFO # Testing TIMER_0:
2018-12-18 15:30:37,802 - INFO # TIMER_0: ERROR on initialization - skipping
2018-12-18 15:30:37,803 - INFO # 
2018-12-18 15:30:37,805 - INFO # 
2018-12-18 15:30:37,817 - INFO # TEST FAILED
```

Are you sure the timer is configured correctly?

**Comment by miri64** [comment]
Also, only squash if the reviewer asks you to. Otherwise it is hard to track your changes + I don't get notified if you force push ;-).

File: boards/common/arduino-atmega/include/periph_conf.h (hunk 8 above)
Comment (tprrt) [reply boards/common/arduino-atmega/include/periph_conf.h]: Indeed, the configuration of the timers was not good, it is corrected.

**Comment by tprrt** [comment]
@miri64 I only saw your comment about the push force, after having pushed my last fix. I'm taking note for next time.

**Comment by tprrt** [comment]
@miri64 I only saw your comment about the push force, after having pushed my last fix. I'm taking note for next time.

**Comment by miri64** [comment]
I'm not sure, when I next come to test this, but it might be that you need to adapt for #9130 if it is merged before :-/

**Comment by kYc0o** [comment]
@miri64 yes it will need some adaptation. However I think is limited to the removal of code duplication, should be straightforward.

**Comment by miri64** [comment]
In the meantime @tprrt: some apps require the `arduino-leonardo` in their `BOARD_INSUFFICIENT_MEMORY` list (see Murdock results). Please add.

**Comment by PeterKietzmann** [comment]
@tprrt the CI didn't pass (have a look at [this output](https://ci.riot-os.org/RIOT-OS/RIOT/7306/60e158a58417e67b9f78e90ffc112f6b62fc5df3/output.html)). On first sight it looks like some applications overflow this platforms memory.

**Comment by miri64** [comment]
I rebased and did that, but sadly I seem to be unable to push to `tprrt:tprrt/leonardo` :(. @tprrt could you allow me to push by setting this tick on the top right under the labels etc.

image

**Comment by miri64** [comment]
Now that #9130 this needs another rebase and some refactoring :-/. @tprrt do you mind if I open a PR based on your changes to speed-up this process?

**Comment by miri64** [comment]
:scream: I did not intend to close this PR.... I somehow managed to push, but now I can't open it anymore oO

**Comment by miri64** [comment]
Since the PR is now closed, I can't re-open the PR >.<...

**Comment by miri64** [comment]
So sorry >.<

**Comment by miri64** [comment]
In the meantime: I stored what I actually wanted to push in https://github.com/miri64/RIOT/tree/leonardo-arduino/feat/initial.

**Comment by keestux** [comment]
> Since the PR is now closed, I can't re-open the PR >.<...
Maybe @tprrt can reopen it. I remember to recently reopen my own
closed PR.

**Comment by maribu** [comment]
In the meantime another round of refactoring of the AVR code took place, so that this PR would need to be rebased and adapted again.

@tprrt: Thank you for the effort you put in this PR. I am under the impression that other helping hand to get this PR merged could be beneficial. If you agree I could create a new PR based on this one and take over from here. (But this would need to wait until I finish my vacation, as I don't have access to the board right now.) (Obviously, the new PR would still pay credit to you via the `@author` tags and copyright headers.)

**Comment by miri64** [comment]
> @tprrt: Thank you for the effort you put in this PR. I am under the impression that other helping hand to get this PR merged could be beneficial. If you agree I could create a new PR based on this one and take over from here. (But this would need to wait until I finish my vacation, as I don't have access to the board right now.) (Obviously, the new PR would still pay credit to you via the `@author` tags and copyright headers.)

@maribu can you maybe do as proposed. You can find the original PR of this PR in the branch I mentioned at https://github.com/RIOT-OS/RIOT/pull/7306#issuecomment-461986657

**Comment by tprrt** [comment]
@maribu No problem, you can do what you said.

**Pull request closed without merging**